File: blind-nav-backend/app/schemas/request.py
from pydantic import BaseModel
from typing import Optional, List


class Note(BaseModel):
    id: str
    text: str
    author: str = "Officer"
    timestamp: str


class IssueRequestCreate(BaseModel):
    transcript: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    category: Optional[str] = None
    priority: Optional[str] = None
    location: Optional[str] = None
    reportedBy: Optional[str] = None
    source: Optional[str] = None


class IssueRequestUpdate(BaseModel):
    status: Optional[str] = None
    assignedTo: Optional[str] = None
    priority: Optional[str] = None
    noteText: Optional[str] = None
    noteAuthor: Optional[str] = "Officer"


class IssueRequest(BaseModel):
    id: str
    title: str
    description: str
    transcript: Optional[str] = None
    category: str
    priority: str
    status: str
    source: str
    location: Optional[str] = None
    reportedBy: Optional[str] = None
    assignedTo: Optional[str] = None
    notes: List[Note] = []
    timestamp: str
    updatedAt: Optional[str] = None

    # ── Phase 2: AI analysis fields ──────────────────────────────────────────
    aiSummary: Optional[str] = None
    aiCategory: Optional[str] = None
    aiPriority: Optional[str] = None
    aiTags: Optional[List[str]] = None
    aiConfidence: Optional[float] = None
    aiReason: Optional[str] = None
    aiEngine: Optional[str] = None
    aiAnalyzedAt: Optional[str] = None
    suggestedAction: Optional[str] = None

    # ── Phase 3: Copilot / agentic workflow fields ───────────────────────────
    copilotSummary: Optional[str] = None
    copilotSuggestedActions: Optional[List[str]] = None
    copilotResolutionSteps: Optional[List[str]] = None
    copilotRecommendedTeam: Optional[str] = None
    copilotRecommendedAssignee: Optional[str] = None
    copilotEscalationLevel: Optional[str] = None      # NORMAL | URGENT | CRITICAL | ESCALATE
    copilotShouldAlert: Optional[bool] = None
    copilotAlertReason: Optional[str] = None
    copilotDraftInternalNote: Optional[str] = None
    copilotDecisionTrace: Optional[List[str]] = None
    copilotEngine: Optional[str] = None               # 'rules' | 'llm'
    copilotGeneratedAt: Optional[str] = None

File: blind-nav-web/src/types/request.ts
export type RequestStatus =
  | 'NEW'
  | 'ASSIGNED'
  | 'IN_PROGRESS'
  | 'COMPLETED'
  | 'CANCELLED';

export type RequestPriority =
  | 'CRITICAL'
  | 'HIGH'
  | 'MEDIUM'
  | 'LOW'
  | 'PENDING_REVIEW';

export type RequestCategory =
  | 'ACCESSIBILITY_SUPPORT'
  | 'PATHWAY_BLOCKED'
  | 'ENTRANCE_BLOCKED'
  | 'LIFT_ISSUE'
  | 'NAVIGATION_AID'
  | 'SAFETY_CONCERN'
  | 'OTHER';

export type AiCategory =
  | 'PHYSICAL_OBSTRUCTION'
  | 'ACCESSIBILITY_EQUIPMENT_ISSUE'
  | 'NAVIGATION_ASSISTANCE'
  | 'UNSAFE_ENVIRONMENT'
  | 'FACILITY_ACCESS_ISSUE'
  | 'EMERGENCY_SUPPORT';

export type CopilotEscalationLevel = 'NORMAL' | 'URGENT' | 'CRITICAL' | 'ESCALATE';

export type RequestSource =
  | 'mobile_voice_report'
  | 'officer_manual_entry'
  | 'assisted_call'
  | 'mobile_app';

export interface Note {
  id: string;
  text: string;
  author: string;
  timestamp: string;
}

export interface IssueRequest {
  id: string;
  title: string;
  description: string;
  transcript?: string;
  category: string;
  priority: string;
  status: string;
  source: string;
  location?: string;
  reportedBy?: string;
  assignedTo?: string;
  notes: Note[];
  timestamp: string;
  updatedAt?: string;

  // ── Phase 2: AI analysis ─────────────────────────────────────────────────
  aiSummary?: string;
  aiCategory?: string;
  aiPriority?: string;
  aiTags?: string[];
  aiConfidence?: number;
  aiReason?: string;
  aiEngine?: string;
  aiAnalyzedAt?: string;
  suggestedAction?: string;

  // ── Phase 4: SLA + Escalation ────────────────────────────────────────────
  slaTargetAt?: string;
  slaStatus?: string;              // ON_TRACK | AT_RISK | BREACHED | COMPLETED
  slaMinutesRemaining?: number;    // negative = overdue
  escalationLevel?: string;        // NONE | TEAM_LEAD | OPERATIONS_MANAGER | EMERGENCY_RESPONSE
  escalatedAt?: string;
  lastEscalationReason?: string;

  // ── Phase 3: Copilot / agentic workflow ──────────────────────────────────
  copilotSummary?: string;
  copilotSuggestedActions?: string[];
  copilotResolutionSteps?: string[];
  copilotRecommendedTeam?: string;
  copilotRecommendedAssignee?: string;
  copilotEscalationLevel?: CopilotEscalationLevel;
  copilotShouldAlert?: boolean;
  copilotAlertReason?: string;
  copilotDraftInternalNote?: string;
  copilotDecisionTrace?: string[];
  copilotEngine?: string;
  copilotGeneratedAt?: string;
}

export interface CreateRequestPayload {
  transcript?: string;
  title?: string;
  description?: string;
  category?: string;
  priority?: string;
  location?: string;
  reportedBy?: string;
  source?: string;
}

export interface UpdateRequestPayload {
  status?: string;
  assignedTo?: string;
  priority?: string;
  noteText?: string;
  noteAuthor?: string;
}

export interface AIAnalysisResult {
  predictedPriority: string;
  predictedCategory: string;
  aiSummary: string;
  tags: string[];
  confidence: number;
  reason: string;
  engine: string;
}

export interface CopilotResult {
  copilotSummary: string;
  copilotSuggestedActions: string[];
  copilotResolutionSteps: string[];
  copilotRecommendedTeam: string;
  copilotRecommendedAssignee: string;
  copilotEscalationLevel: CopilotEscalationLevel;
  copilotShouldAlert: boolean;
  copilotAlertReason: string;
  copilotDraftInternalNote: string;
  copilotDecisionTrace: string[];
  copilotEngine: string;
  copilotGeneratedAt?: string;
}
